03-rbac/proposal.md

```diff
@@ -0,0 +1,186 @@
+# Summary
+
+This proposal is meant to start a discussion about Role Based Access Control (or RBAC for short). A number of users have been asking for ways to restrict the permissions given to various users in the system, so this will be an attempt to outline what these restrictions might look like, and how they might work.
+
+One of the main drawbacks of implementing RBAC is that there can be performance implications. These typically come about when we need to make expensive queries to determine wether or not a user can perform the intended action. This isn't a good enough reason not to implement such checks, but we should be mindful of the potential consequences.
+
+We'll try to answer a few questions about how this might work. Specifically:
+
+- What roles do we need?
+- How would fly set-team work?
+- How do roles get persisted in the database?
+- How are roles handled by the API?
+
+
+# Proposal
+
+NOTE: We've followed up with a ✅ beside each of the options that we've chosen below.
+
+
+## What roles do we need?
+
+`admin` - Granted if you are part of an admin team. Currently only the `main` team is an admin team.
+
+`owner` - Granted during `fly set-team`. Has read/write access to all the team's resources, as well as admin priviledges for the team itself (`set-team`, `rename-team`, etc).
+
+`member` - Granted during `fly set-team`. Has read/write access to all the team's resources.
+
+`viewer` - Granted during `fly set-team`. Has readonly access to all the team's resources.
+
+
+## How would fly set-team work?
+
+Fly set-team should work the same as before:
+```bash
+fly -t mytarget set-team -n myteam --allow-all-users
+fly -t mytarget set-team -n myotherteam --github-user pivotal-jwinters --github-team myorg:myteam
+```
+
+This will assign the `owner` role to the specified users and groups. If you want more control over which roles get assigned, you can provide a config file with the mappings.
+
+```bash
+fly -t mytarget set-team -n myteam -c /tmp/team-config
+```
+
+Where `/tmp/team-config` might look something like:
+
+### Option 1 - roles array
+
+```yaml
+roles: 
+- name: owner
+  local: 
+    users: ["some-admin"]
+- name: member
+  github: 
+    users: ["my-github-login"]
+    teams: ["my-org:my-github-team"]
+  cf: 
+    users: ["myusername"]
+    spaces: ["myorg:myspace"]
+- name: viewer
+  allow_all_users: true
+```
+
+or this:
+
+### ✅ Option 2 - roles map
```

Comment:
Just want to say that I like this quite a bit. Git-backed configuration is incredibly powerful, and it allows self-service for Concourse access: it allows teams within the company to create PRs against the config repo, and then CI can make the changes once things get merged into master.